### PR TITLE
[parser] Fix indentation tracking after line continuations

### LINKED ISSUE #19301: Syntax errors related to indentation after backslashes
### Summary

Ruff has false positives and false negatives when detecting syntax errors related to indentation after backslashes. [Python’s documentation](https://docs.python.org/3/reference/lexical_analysis.html#indentation) says “the whitespace up to the first backslash determines the indentation”, so the leading white space of the next line shouldn’t cause a syntax error, though this principle has been interpreted inconsistently in different Python versions. See https://github.com/python/cpython/issues/90249.

[False positive (unexpected indentation)](https://play.ruff.rs/44ebd0a2-e83f-4898-8f31-44d0c631598f) (in all Python versions):
```console
$ cat >test_1.py <<'# EOF'
if True:
    pass
    \
        print("1")
# EOF

$ python3.13 test_1.py
1

$ ruff --isolated check test_1.py --output-format concise -q --target-version py313
test_1.py:3:1: SyntaxError: Unexpected indentation
test_1.py:5:1: SyntaxError: Expected a statement
```

[False positive (missing indentation)](https://play.ruff.rs/029a16c9-bfa6-4ca2-9597-26c3036fdd23) (except in Python 3.9 only, where it is a true positive):
```console
$ cat >test_2.py <<'# EOF'
if True:
    \
print("2")
# EOF

$ python3.13 test_2.py
2

$ ruff --isolated check test_2.py --output-format concise -q --target-version py313
test_2.py:3:1: SyntaxError: Expected an indented block after `if` statement

$ python3.9 test_2.py 2>&1 | tail -n 1
IndentationError: expected an indented block

$ ruff --isolated check test_2.py --output-format concise -q --target-version py39
test_2.py:3:1: SyntaxError: Expected an indented block after `if` statement
```

[False negative (unexpected indentation)](https://play.ruff.rs/b085d040-c518-45ab-a8df-94688c241ebf) (except in Python 3.9 and earlier, where it is a true negative):
```console
$ cat >test_3.py <<'# EOF'
\
    \
print("3")
# EOF

$ python3.13 test_3.py 2>&1 | tail -n 1
IndentationError: unexpected indent

$ ruff --isolated check test_3.py --target-version py313
All checks passed!

$ python3.9 test_3.py
3

$ ruff --isolated check test_3.py --target-version py39
All checks passed!
```

[False positive (unexpected indentation)](https://play.ruff.rs/c4359afd-c1e0-4dd5-b2bc-cc9e79857903) (except in Python 3.10 and later, where it is a true positive):
```console
$ cat >test_4.py <<'# EOF'
\
    print("4")
# EOF

$ python3.9 test_4.py
4

$ ruff --isolated check test_4.py --output-format concise -q --target-version py39
test_4.py:1:1: SyntaxError: Unexpected indentation
test_4.py:3:1: SyntaxError: Expected a statement

$ python3.10 test_4.py 2>&1 | tail -n 1
IndentationError: unexpected indent

$ ruff --isolated check test_4.py --output-format concise -q --target-version py310
test_4.py:1:1: SyntaxError: Unexpected indentation
test_4.py:3:1: SyntaxError: Expected a statement
```

[False positive (no matching indentation)](https://play.ruff.rs/51876800-57da-4c30-99cb-84a0a2f5b1a8) (in all Python versions):
```console
$ cat >test_5.py <<'# EOF'
if True:
    pass
    \
  print("5")
# EOF

$ python3.13 test_5.py
5

$ ruff --isolated check test_5.py --output-format concise -q --target-version py313
test_5.py:3:1: SyntaxError: unindent does not match any outer indentation level
test_5.py:5:1: SyntaxError: Expected dedent, found end of file
```

### Version

ruff 0.12.3 (5bc81f26c 2025-07-11)

### PR DESCRIPTION
**Pull request by kar-ganap**
## Summary

Fixes #19301.

Per the [Python spec](https://docs.python.org/3/reference/lexical_analysis.html#indentation), the whitespace up to the first backslash determines the line's indentation, and continuation lines should not affect the indentation level. The lexer's \`eat_indentation\` loop was not handling this: after a \`\\\` continuation, it accumulated the continuation line's whitespace into the indentation, causing spurious \`IndentationError\` and "Expected a statement" errors.

### Reproduction

\`\`\`python
if True:
    pass
    \\
        print("1")
\`\`\`

**Before:** \`SyntaxError: Unexpected indentation\` + \`Expected a statement\`
**After:** Parses correctly (matches CPython behavior)

### Fix

After consuming \`\\<newline>\` during indentation tracking in \`eat_indentation\`, skip the continuation line's whitespace with \`eat_while(is_python_whitespace)\` without accumulating it into \`indentation\`. Guard: only skip when \`indentation != Indentation::root()\` — when \`\\\` is at column 0, let the loop continue so the next line's whitespace is accumulated normally (needed for the \`else: \\\` pattern in RET505 fixtures).

Reference

- Python spec: https://docs.python.org/3/reference/lexical_analysis.html#indentation
- CPython discussion: https://github.com/python/cpython/issues/90249

## Test plan

Add lexer test cases, a couple of inline parser tests.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
> Per the Python spec, the whitespace up to the first backslash determines the line's indentation, and continuation lines should not affect the indentation level.

Can you provide a reference to where this is specified? It'd be useful to add a link to it in the comments.

**Comment by dhruvmanila** [comment]
> Reproduction from the issue:

This example does not raise an error on latest `main`: https://play.ruff.rs/ccc71f1d-f784-4b53-b253-0d220a765fd4 although the test case added does raise a syntax error.

**Comment by kar-ganap** [comment]
@dhruvmanila Thanks for the review!

**Re: Python spec reference** — The reference is in the [Python Language Reference — Indentation](https://docs.python.org/3/reference/lexical_analysis.html#indentation). The original issue also links to the related CPython discussion: https://github.com/python/cpython/issues/90249. I've updated the PR description to include the spec link.

**Re: reproduction not erroring on main** — Good catch, the reproduction in the original PR description was wrong (it showed a mid-expression `x = \` continuation which is handled differently). The actual bug is when `\` appears at the start of a line during indentation tracking, like the issue's test_1:

```python
if True:
    pass
    \
        print("1")
```

This still errors on latest main:
```
$ ruff check test.py --isolated --output-format concise -q
test.py:3:1: SyntaxError: Unexpected indentation
test.py:5:1: SyntaxError: Expected a statement
```

I've updated the PR description with the correct reproduction. The regression test (`backslash_continuation_indentation.py`) tests this exact case and was correctly flagging the bug.

File: crates/ruff_python_parser/src/lexer.rs [hunk 1]
```diff
@@ -263,7 +263,31 @@ impl<'src> Lexer<'src> {
                             self.token_range(),
                         )));
                     }
-                    indentation = Indentation::root();
+                    // test_ok backslash_continuation_indentation
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/lexer.rs:266]: Lexer tests are written using Rust code under the `#[cfg(test)]` section in this file which would create snapshots of the lexed tokens. Can you move this down below and add a few more edge cases?

File: crates/ruff_python_parser/src/lexer.rs [hunk 2]
```diff
@@ -263,7 +263,31 @@ impl<'src> Lexer<'src> {
...
+                    // continuation-line whitespace without accumulating it
+                    // into `indentation`. However, if the backslash is at
+                    // column 0 (no prior indentation), let the loop continue
+                    // so the next line's whitespace is accumulated normally.
+                    //
+                    // [1]: https://docs.python.org/3/reference/lexical_analysis.html#indentation
+                    // See also: https://github.com/python/cpython/issues/90249
+                    if indentation != Indentation::root() {
+                        self.cursor
+                            .eat_while(|c| c == ' ' || c == '\t' || c == '\x0C');
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/lexer.rs]: We should reuse `is_python_whitespace` here

File: crates/ruff_python_parser/src/lexer.rs [hunk 3]
```diff
@@ -263,7 +263,31 @@ impl<'src> Lexer<'src> {
...
+                    // Per Python spec [1], the whitespace up to the first
+                    // backslash determines the line's indentation. Skip
+                    // continuation-line whitespace without accumulating it
+                    // into `indentation`. However, if the backslash is at
+                    // column 0 (no prior indentation), let the loop continue
+                    // so the next line's whitespace is accumulated normally.
+                    //
+                    // [1]: https://docs.python.org/3/reference/lexical_analysis.html#indentation
+                    // See also: https://github.com/python/cpython/issues/90249
+                    if indentation != Indentation::root() {
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/lexer.rs:299]: Can you also add a test case when this condition is true but it's still a syntax error?

File: crates/ruff_python_parser/src/lexer.rs [hunk 4]
```diff
@@ -263,7 +263,31 @@ impl<'src> Lexer<'src> {
                             self.token_range(),
                         )));
                     }
-                    indentation = Indentation::root();
+                    // test_ok backslash_continuation_indentation
+                    // if True:
+                    //     pass
+                    //     \
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/lexer.rs:269]: What happens when there are multiple backslashes? The spec says "the whitespace up to the _first_ backslash determines the indentation" but there could be more, right?

File: crates/ruff_python_parser/src/lexer.rs (hunk 1 above)
Comment (kar-ganap) [reply crates/ruff_python_parser/src/lexer.rs:266]: Done — added 12 lexer snapshot tests in the `#[cfg(test)]` section (4 scenarios × 3 EOL variants), following the existing helper + per-EOL pattern. Kept the inline tests as documentation for the fixture infrastructure.

File: crates/ruff_python_parser/src/lexer.rs (hunk 2 above)
Comment (kar-ganap) [reply crates/ruff_python_parser/src/lexer.rs]: Done — replaced with `eat_while(is_python_whitespace)`.

File: crates/ruff_python_parser/src/lexer.rs (hunk 3 above)
Comment (kar-ganap) [reply crates/ruff_python_parser/src/lexer.rs:299]: Added `backslash_continuation_dedent_error` test: `def f(): pass` followed by `  \\` (2-space indent + continuation). The indent stack is [0, 4] but the indentation is 2, producing an `IndentationError` at the lexer level.

File: crates/ruff_python_parser/src/lexer.rs (hunk 4 above)
Comment (kar-ganap) [reply crates/ruff_python_parser/src/lexer.rs:269]: Each `\\` in the `eat_indentation` loop is processed independently. After eating `\\<newline>`, `eat_while(is_python_whitespace)` consumes the continuation line's whitespace. The loop continues — if the next char is another `\\`, the process repeats. The `indentation` variable is never modified by the `\\` arm, so it retains the value from before the first `\\`. The new `backslash_continuation_multiple` snapshot test demonstrates this across all 3 EOL types.

File: crates/ruff_python_parser/src/lexer.rs [hunk 5]
```diff
@@ -3061,4 +3096,95 @@ t"{(lambda x:{x})}"
             UnterminatedTripleQuotedString
         );
     }
+
+    // Backslash at indented position: continuation whitespace should not
+    // affect indentation level.
+    fn backslash_continuation_indentation_eol(eol: &str) -> LexerOutput {
+        let source = format!("if True:{eol}    pass{eol}    \\{eol}        print(\"1\"){eol}");
+        lex_source(&source)
+    }
```
Comment (dhruvmanila) [inline crates/ruff_python_parser/src/lexer.rs]: Sorry but this is not what I meant, we don't need to test for various EOL cases as that's already tested, the test cases should be specific to the behavior that's being changed.

I've updated the test cases locally and I've found that the following isn't raising a syntax error when it should:
```py
if True:
    1
      \
    2
```

The indentation level at `1` is of 4 spaces while at `2` is of 6 spaces. So, I think there's more that needs to be done. I need to take a closer look at the solution and what CPython is doing.

Comment (dhruvmanila) [reply crates/ruff_python_parser/src/lexer.rs]: I'll look at it a later today unless you can figure out what's wrong, I've put the PR into draft for the time being.

Comment (kar-ganap) [reply crates/ruff_python_parser/src/lexer.rs]: I investigated the overindented case you raised:

```python
if True:
    1
      \
    2
```

This **does** produce an error with our fix. The 6 spaces before `\` freeze the indentation at 6, so the lexer emits an `Indent` from 4→6, and the parser rejects it as "Unexpected indentation":

```
invalid-syntax: Unexpected indentation
 --> test.py:3:1
  |
1 |   if True:
2 |       1
3 | /       \
4 | |     2
  | |____^
```

The issue exists **on main** (without this fix), where `indentation = Indentation::root()` resets indentation to 0, then re-accumulates 4 spaces from the continuation line — matching the block level and silently accepting the overindent. Our fix corrects this.

I've added a `backslash_continuation_overindented` snapshot test that captures the token stream for this exact case (showing the two `Indent` tokens). Also simplified the tests per your feedback — removed the 12 EOL-variant tests and replaced them with 5 targeted tests:

1. `backslash_continuation_in_block` — basic valid case
2. `backslash_continuation_at_column_zero` — column 0 special case  
3. `backslash_continuation_multiple` — consecutive continuations
4. `backslash_continuation_overindented` — your case (indentation frozen at 6, Indent 4→6 emitted)
5. `backslash_continuation_mismatch` — indentation doesn't match stack, lexer error

Verified all cases match CPython behavior.

Comment (dhruvmanila) [reply crates/ruff_python_parser/src/lexer.rs]: I'm sure you didn't mean to do that but you force pushed some of the changes which removed my changes in https://github.com/astral-sh/ruff/commit/879255dae4b5ae98dfaf12c23980df1fb33c43cf. Please avoid making any further changes to avoid any conflict.

**Comment by dhruvmanila** [comment]
Closing in favor of https://github.com/astral-sh/ruff/pull/23417
